internal/runbits/buildlogfile/buildlogfile.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/platform/runtime/artifact"
+	"github.com/ActiveState/cli/pkg/platform/runtime/setup/buildlog"
+)
+
+// maxConcurrency is the maximum number of artifact-build-logs that we download in parallel in order to add the information to the build log file.
+const maxConcurrency = 5
+
+type BuildLogFile struct {
+	out              output.Outputer
+	logFile          *os.File
+	numInstalled     int
+	numToBeInstalled int
+	ctx              context.Context
+	cancel           context.CancelFunc
+	logsCh           chan artifactLog
+}
+
+type artifactLog struct {
+	artifactID     artifact.ArtifactID
+	artifactName   string
+	unsignedLogURI string
+}
+
+// verboseLogging returns true if the user provided an environment variable for it
+func verboseLogging() bool {
+	return os.Getenv(constants.LogBuildVerboseEnvVarName) == "true"
+}
+
+// downloadAndAddArtifactLog downloads an artifact build log and adds it to the build log file
+func (bl *BuildLogFile) downloadAndAddArtifactLog(ul artifactLog, ctx context.Context) error {
+	unsignedURL, err := url.Parse(ul.unsignedLogURI)
+	if err != nil {
+		return errs.Wrap(err, "Could not parse log URL %s", ul.unsignedLogURI)
+	}
+	logURL, err := model.SignS3URL(unsignedURL)
+	if err != nil {
+		return errs.Wrap(err, "Could not sign log url %s", unsignedURL)
+	}
+
+	// download the log and stream it line-by-line
+	logging.Debug("downloading logURI: %s", logURL.String())
+	req, err := http.NewRequestWithContext(ctx, "GET", logURL.String(), nil)
+	if err != nil {
+		return errs.Wrap(err, "Failed to create GET request for logURL.")
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errs.Wrap(err, "Failed to execute HTTP GET request for logURL.")
+	}
+
+	defer resp.Body.Close()
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		// we need to unmarshal every line
+		var am buildlog.ArtifactProgressMessage
+		if err := json.Unmarshal(line, &am); err != nil {
+			return errs.Wrap(err, "Failed to unmarshal build log line")
+		}
+		if err := bl.BuildArtifactProgress(ul.artifactID, ul.artifactName, am.Timestamp, am.Body.Message, am.Body.Facility, am.PipeName, am.Source); err != nil {
+			return errs.Wrap(err, "Failed to write build log line")
+		}
+	}
+	return nil
+}
+
+func New(out output.Outputer) (*BuildLogFile, error) {
+	logFile, err := os.CreateTemp("", fmt.Sprintf("build-log-%s", time.Now().Format("060102030405")))
+	if err != nil {
+		return nil, errs.Wrap(err, "Failed to create temporary build log file")
+	}
+
+	bl := &BuildLogFile{out: out, logFile: logFile}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+	bl.logsCh = make(chan artifactLog)
+	for i := 0; i < maxConcurrency; i++ {
```

Comment:
Might be more appropriate/simpler to use workerpool here, eg:

http://github.com/ActiveState/cli/blob/3dc202147f99ba6f4a0711e9baf11961dc725f90/pkg/platform/runtime/setup/setup.go#L328-L328

Reply:
I see your point.  This can indeed block (in the consumer loop) in case we are downloading a lot of build logs, and if the download is hanging. I just wanted to limit the number of concurrent go functions here.

Reply:
I think I know how to solve this: I will drop sending requests (in `addBuildLogs` if the request takes more than one second.)  We can log that to rollbar.  If we are getting to many rollbar errors, we can increase the `maxConcurrency` number.